File: LeetCode/Python Solutions/Perfect Squares/perfect.py
class Solution:
    def numSquares(self, n: int) -> int:
        # DP array initialized with infinity for all indexes
        dp = [float('inf')] * (n + 1)
        
        # Base case: 0 can be represented by 0 squares
        dp[0] = 0
        
        # Iterate through all numbers from 1 to n to fill the dp array
        for i in range(1, n + 1):
            # Try all square numbers less than i
            j = 1
            while j * j <= i:
                dp[i] = min(dp[i], dp[i - j * j] + 1)
                j += 1
        
        return dp[n]

File: LeetCode/Java Solutions/Perfect Squares/perfect.java
public class Solution {
    public int numSquares(int n) {
        int[] dp = new int[n + 1];
        
        // Fill the array with the maximum value
        for (int i = 0; i <= n; i++) {
            dp[i] = Integer.MAX_VALUE;
        }
        
        // Base case
        dp[0] = 0;
        
        // Build up the dp table
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j * j <= i; j++) {
                dp[i] = Math.min(dp[i], dp[i - j * j] + 1);
            }
        }
        
        return dp[n];
    }
}
